stdlib/public/runtime/StringComparison.cpp

```diff
@@ -0,0 +1,162 @@
+//===--- StringComparison.cpp ---------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Runtime/StringComparison.h"
+#include "swift/Runtime/Config.h"
+#include <cassert>
+#include <cstdint>
+
+#ifdef __APPLE__
+
+#include <simd/simd.h>
+
+static int findDiffIdxSmall(uint8_t *left, uint16_t *right, int count);
+
+namespace {
+static constexpr simd_packed_uchar16 zero = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+                                             0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+                                             0x0, 0x0, 0x0, 0x0};
+} // end anonymous namespace
+
+/// This is equivalent to _mm_unpacklo_epi8(x, zero) but written in a manner
+/// that should work across platform.
+static simd_packed_ushort8 unpackLower(simd_packed_uchar16 input) {
+  return (simd_packed_ushort8)__builtin_shufflevector(
+      input, zero, 0, 16, 1, 17, 2, 18, 3, 19, 4, 20, 5, 21, 6, 22, 7, 23);
+}
+
+/// This is equivalent to _mm_unpackhi_epi8(x, zero) but written in a manner
+/// that should work across platform.
+static simd_packed_ushort8 unpackUpper(simd_packed_uchar16 input) {
+  return (simd_packed_ushort8)__builtin_shufflevector(input, zero, 8, 24, 9, 25,
+                                                      10, 26, 11, 27, 12, 28,
+                                                      13, 29, 14, 30, 15, 31);
+}
+
+/// This is equivalent to _mm_unpacklo_epi8(lhs, zero) == rhs
+static simd_packed_short8 performComparisonLow(simd_packed_uchar16 lhs,
```

Comment:
Instead of widening UInt8 to UInt16 for the comparison, I would do:
~~~~
simd::char16 equalP(simd::ushort16 lhs, simd::uchar16 rhs) {
  auto lhsAsUInt8 = (simd::uchar32)lhs;
  // lhs == rhs if the low bytes are equal and the high-byte of lhs is zero.
  return lhsAsUInt8.even == rhs & lhsAsUInt8.odd == 0;
}
~~~~

Reply:
I checked the assembly. It doesn't look as good. See:

https://gist.github.com/gottesmm/08b5d69e7bf1dc2a83d75b1b592dec45#file-test-s-L18
vs
https://gist.github.com/gottesmm/08b5d69e7bf1dc2a83d75b1b592dec45#file-test-s-L81

Reply:
Well, that's a codegen bug in LLVM that we should fix =)